crates/biome_js_analyze/src/lint/nursery/no_useless_undefined.rs

```diff
@@ -0,0 +1,447 @@
+        | JsCallExpression
+        | JsArrowFunctionExpression
+        | JsFormalParameter
+}
+
+static FUNCTION_NAMES: [&str; 28] = [
+    // Function#bind()
+    "bind",
+    // Compare function names
+    "is",
+    "equal",
+    "notEqual",
+    "strictEqual",
+    "notStrictEqual",
+    "propertyVal",
+    "notPropertyVal",
+    "not",
+    "include",
+    "property",
+    "toBe",
+    "toHaveBeenCalledWith",
+    "toContain",
+    "toContainEqual",
+    "toEqual",
+    "same",
+    "notSame",
+    "strictSame",
+    "strictNotSame",
+    // `array.push(undefined)`
+    "push",
+    // `array.unshift(undefined)`
+    "unshift",
+    // `array.includes(undefined)`
+    "includes",
+    // `set.add(undefined)`
+    "add",
+    // `set.has(undefined)`
+    "has",
+    // `map.set(foo, undefined)`
+    "set",
+    // `React.createContext(undefined)`
+    "createContext",
+    // https://vuejs.org/api/reactivity-core.html#ref
+    "ref",
+];
+
+fn should_ignore_function(expr: &AnyJsExpression) -> bool {
+    let name = match expr {
+        AnyJsExpression::JsIdentifierExpression(ident) => ident.text(),
+        AnyJsExpression::JsStaticMemberExpression(member_expr) => {
+            let member = member_expr.member().ok().unwrap();
+            member.text()
+        }
+        _ => return false,
+    };
+
+    FUNCTION_NAMES.contains(&name.as_str()) ||
+    // setState(undefined), setXXX(undefined)
+    name.starts_with("set")
+}
+
+fn is_undefined(expr: Option<&AnyJsExpression>) -> Option<(String, TextRange)> {
+    let ident = expr?.as_js_reference_identifier()?;
+    if ident.is_undefined() {
+        Some((ident.text().to_string(), ident.range()))
+    } else {
+        None
+    }
+}
+
+impl Rule for NoUselessUndefined {
+    type Query = Ast<RuleQuery>;
+    type State = (String, TextRange);
+    type Signals = Vec<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let mut signals = vec![];
+
+        match node {
+            // let foo = undefined;
+            RuleQuery::JsVariableStatement(statement) => {
+                let Ok(node) = statement.declaration() else {
+                    return signals;
+                };
+                let let_or_var = node.is_let() || node.is_var();
+                if !let_or_var {
+                    return signals;
+                }
+
+                for declarator in node.declarators() {
+                    let Ok(decl) = declarator else { continue };
+                    if let Some(initializer) = decl.initializer() {
+                        let expr = initializer.expression().ok();
+                        if let Some(state) = is_undefined(expr.as_ref()) {
+                            signals.push(state);
+                        }
+                    }
+                }
+            }
+            // foo(bar, undefined, undefined);
+            RuleQuery::JsCallExpression(js_call_expr) => {
```

Comment:
I wonder if we should not completely ignore setting function calls. There are many cases where it makes the code more readable as noted by the exceptions we have for `includes` and others.
Any opinion?

Reply:
That's good. Certainly when I write TypeScript myself, I want to explicitly indicate undefined for function arguments.

Reply:
Let's ignore function calls :)